src/python/pants/backend/java/package/fat_jar.py

```diff
@@ -0,0 +1,241 @@
+    #
+
+    main_class = field_set.main_class.value
+
+    manifest_content = FileContent(
+        "META-INF/MANIFEST.MF",
+        # NB(chrisjrn): we're joining strings with newlines, becuase the JAR manfiest format
+        # needs precise indentation, and _cannot_ start with a blank line. `dedent` seriously
+        # messes up those requirements.
+        "\n".join(
+            [
+                "Manifest-Version: 1.0",
+                f"Main-Class: {main_class}",
+                "",  # THIS BLANK LINE WILL BREAK EVERYTHING IF DELETED. DON'T DELETE IT.
+            ]
+        ).encode("utf-8"),
+    )
+
+    manifest_digest = await Get(Digest, CreateDigest([manifest_content]))
+    thin_jar_inputs = await Get(
+        Digest, MergeDigests([compiled_class_files_digest, manifest_digest])
+    )
+
+    thin_jar_result = await Get(
+        ProcessResult,
+        Process(
+            argv=[
+                "/usr/bin/zip",
+                "pants_thin_jar.jar",
+                *compiled_class_files_snapshot.files + ("META-INF/MANIFEST.MF",),
+            ],
+            description="Build thin JAR file.",
+            input_digest=thin_jar_inputs,
+            output_files=["pants_thin_jar.jar"],
+        ),
+    )
+
+    logger.info(thin_jar_result.stdout.decode())
+    logger.info(thin_jar_result.stderr.decode())
+
+    thin_jar = thin_jar_result.output_digest
+
+    #
+    # 3. Create broken fat JAR
+    #
+
+    lockfile_requests = [
+        Get(
+            CoursierResolvedLockfile,
+            CoursierLockfileForTargetRequest(
+                Targets(CompileJavaSourceRequest(tgt).component.members)
+            ),
+        )
+        for tgt in sources_to_compile
+    ]
+    lockfiles = await MultiGet(lockfile_requests)
+
+    materialized_classpath = await Get(
+        MaterializedClasspath,
+        MaterializedClasspathRequest(
+            prefix="coursier",
+            lockfiles=lockfiles,
+        ),
+    )
+
+    # NB(chrisjrn) Concatenating multiple ZIP files produces a zip file that is _mostly_ safe to
+    # be distributed (it can be fixed with `-FF`), so that's how we construct our fat JAR
+    # without exploding the files to disk.
+    #
+    # `ZIP` files are extracted top-to-bottom and archives can have duplicate names
+    # (e.g. `META-INF/MANIFEST.MF`). In the case of a `JAR` file, the JVM will understand the 
+    # last file with that file name to be the actual one. Therefore, our thin JAR needs to be
+    # appear at the end of the file for (in particular) our manifest to take precedence.
+    # If there are duplicate classnames at a given package address fro JARs, then 
+    # behaviour will be non-deterministic. Sorry!
+    cat_script = FileContent(
+        "_cat_zip_files.sh",
+        textwrap.dedent(
+            f"""
+            /bin/cat {" ".join(materialized_classpath._reified_filenames())} pants_thin_jar.jar > pants_broken_fat_jar.jar
+            """
+        ).encode("utf-8"),
```

Comment:
You'll want to locate `cat` as well. But it's also generally a bit more compact to use an inline bash script with `bash -c`: https://github.com/pantsbuild/pants/blob/52c0a3be635511a37fda5c10a291e9fd59b2eea9/src/python/pants/backend/java/util_rules.py#L66-L70